lib/core/src/chain_swap.rs

```diff
@@ -681,148 +723,298 @@ impl ChainSwapStateHandler {
+        };
+        BoltzApiClientV2::new(boltz_url)
+    }
+
+    pub async fn prepare_refund(
         &self,
         lockup_address: &str,
-        output_address: &str,
-        sat_per_vbyte: u32,
+        refund_address: &str,
+        fee_rate_msat_per_vb: u32,
     ) -> SdkResult<(u32, u64, Option<String>)> {
         let swap = self
             .persister
             .fetch_chain_swap_by_lockup_address(lockup_address)?
             .ok_or(SdkError::Generic {
                 err: format!("Swap {} not found", lockup_address),
             })?;
-        if let Some(refund_tx_id) = swap.refund_tx_id.clone() {
+
+        let refund_tx_id = swap.refund_tx_id.clone();
+        if let Some(refund_tx_id) = &refund_tx_id {
             warn!(
                 "A refund tx for Chain Swap {} was already broadcast: txid {refund_tx_id}",
                 swap.id
             );
         }
-        let (tx_vsize, tx_fee_sat) =
-            self.swapper
-                .prepare_chain_swap_refund(&swap, output_address, sat_per_vbyte as f32)?;
-        Ok((tx_vsize, tx_fee_sat, swap.refund_tx_id))
+
+        let (refund_tx_size, refund_tx_fees_sat) = self.swapper.estimate_refund_broadcast(
+            Swap::Chain(swap),
+            refund_address,
+            Some(fee_rate_msat_per_vb as f64 / 1000.0),
+        )?;
+
+        Ok((refund_tx_size, refund_tx_fees_sat, refund_tx_id))
     }
 
     pub(crate) async fn refund_incoming_swap(
         &self,
         lockup_address: &str,
-        output_address: &str,
-        sat_per_vbyte: u32,
+        refund_address: &str,
+        broadcast_fee_rate_msat_per_vb: u32,
+        is_cooperative: bool,
     ) -> Result<String, PaymentError> {
         let swap = self
             .persister
             .fetch_chain_swap_by_lockup_address(lockup_address)?
             .ok_or(PaymentError::Generic {
                 err: format!("Swap {} not found", lockup_address),
             })?;
-        if let Some(refund_tx_id) = swap.refund_tx_id.clone() {
-            warn!(
-                "A refund tx for Chain Swap {} was already broadcast: txid {refund_tx_id}",
-                swap.id
-            );
-        }
 
-        let (_, broadcast_fees_sat) =
-            self.swapper
-                .prepare_chain_swap_refund(&swap, output_address, sat_per_vbyte as f32)?;
-        let refund_res =
-            self.swapper
-                .refund_chain_swap_cooperative(&swap, output_address, broadcast_fees_sat);
-        let refund_tx_id = match refund_res {
-            Ok(res) => Ok(res),
-            Err(e) => {
-                warn!("Cooperative refund failed: {:?}", e);
-                let current_height = self.bitcoin_chain_service.lock().await.tip()?.height as u32;
-                self.swapper.refund_chain_swap_non_cooperative(
-                    &swap,
-                    broadcast_fees_sat,
-                    output_address,
-                    current_height,
+        ensure_sdk!(
```

Comment:
Shouldn't we also ensure the swap is refundable? to prevent broadcasting an invalid transaction?

Reply:
Agreed